File: src/sql/postgres/Signature.zig
const Signature = @This();

fields: []const int4,
name: []const u8,
query: []const u8,
prepared_statement_name: []const u8,

pub fn empty() Signature {
    return Signature{
        .fields = &[_]int4{},
        .name = &[_]u8{},
        .query = &[_]u8{},
        .prepared_statement_name = &[_]u8{},
    };
}

pub fn deinit(this: *Signature) void {
    if (this.prepared_statement_name.len > 0) {
        bun.default_allocator.free(this.prepared_statement_name);
    }
    if (this.name.len > 0) {
        bun.default_allocator.free(this.name);
    }
    if (this.fields.len > 0) {
        bun.default_allocator.free(this.fields);
    }
    if (this.query.len > 0) {
        bun.default_allocator.free(this.query);
    }
}

pub fn hash(this: *const Signature) u64 {
    var hasher = std.hash.Wyhash.init(0);
    hasher.update(this.name);
    hasher.update(std.mem.sliceAsBytes(this.fields));
    return hasher.final();
}

pub fn generate(globalObject: *jsc.JSGlobalObject, query: []const u8, array_value: JSValue, columns: JSValue, prepared_statement_id: u64, unnamed: bool) !Signature {
    var fields = std.array_list.Managed(int4).init(bun.default_allocator);

    errdefer {
        fields.deinit();
    }

    var iter = try QueryBindingIterator.init(array_value, columns, globalObject);

    while (try iter.next()) |value| {
        if (value.isEmptyOrUndefinedOrNull()) {
            // Allow postgres to decide the type
            try fields.append(0);
            continue;
        }

        const tag = try types.Tag.fromJS(globalObject, value);

        switch (tag) {
            .bool, .int4, .int8, .float8, .int2, .numeric, .float4, .bytea => {
                // We decide the type
                try fields.append(@intFromEnum(tag));
            },
            else => {
                // Allow postgres to decide the type
                try fields.append(0);
            },
        }
    }

    if (iter.anyFailed()) {
        return error.InvalidQueryBinding;
    }

    // The statement cache key (`signature.name`) is just the SQL text — do not
    // depend on the runtime null/type pattern of the bound values. Otherwise
    // each distinct null pattern in a bulk insert allocates a fresh server-side
    // prepared statement and leaks memory on the database for the life of the
    // connection. See issue #28980.
    //
    // Subsequent `Bind` messages carry the per-row null flag (-1 length) in
    // the wire format, and `statement.parameters` (populated from the server's
    // `ParameterDescription`) is what picks the binary/text format on re-bind,
    // so the parameter-type OIDs sent in the initial `Parse` do not need to
    // be re-keyed per null pattern.
    const name = try bun.default_allocator.dupe(u8, query);
    errdefer bun.default_allocator.free(name);

    // max u64 length is 20, max prepared_statement_name length is 63
    const prepared_statement_name = if (unnamed) "" else try std.fmt.allocPrint(bun.default_allocator, "P{s}${d}", .{ name[0..@min(40, name.len)], prepared_statement_id });

    return Signature{
        .prepared_statement_name = prepared_statement_name,
        .name = name,
        .fields = fields.items,
        .query = try bun.default_allocator.dupe(u8, query),
    };
}

const bun = @import("bun");
const std = @import("std");
const QueryBindingIterator = @import("../shared/QueryBindingIterator.zig").QueryBindingIterator;

const types = @import("./PostgresTypes.zig");
const int4 = types.int4;

const jsc = bun.jsc;
const JSValue = jsc.JSValue;

File: src/sql_jsc/postgres/Signature.rs
use crate::jsc::{JSGlobalObject, JSValue};
use bun_sql::postgres::postgres_types::Int4;

#[derive(Default)]
pub struct Signature {
    pub fields: Box<[Int4]>,
    pub name: Box<[u8]>,
    pub query: Box<[u8]>,
    pub prepared_statement_name: Box<[u8]>,
}

impl Signature {
    pub fn empty() -> Signature {
        Signature {
            fields: Box::default(),
            name: Box::default(),
            query: Box::default(),
            prepared_statement_name: Box::default(),
        }
    }

    // No explicit `Drop` impl needed: the `Box<[T]>` fields free the four owned slices automatically.

    // JSError (from QueryBindingIterator /
    // Tag::from_js), OOM, and InvalidQueryBinding are collapsed to the
    // crate-wide `crate::Error`.
    pub fn generate(
        global_object: &JSGlobalObject,
        query: &[u8],
        array_value: JSValue,
        columns: JSValue,
        prepared_statement_id: u64,
        unnamed: bool,
    ) -> crate::Result<Signature> {
        use crate::jsc::js_error_to_postgres;
        use crate::postgres::types::tag_jsc;
        use crate::shared::QueryBindingIterator;
        use bun_sql::postgres::types::tag::Tag;

        let mut fields: Vec<Int4> = Vec::new();
        let mut name: Vec<u8> = Vec::with_capacity(query.len());

        name.extend_from_slice(query);

        let mut iter = QueryBindingIterator::init(array_value, columns, global_object)
            .map_err(js_error_to_postgres)?;

        while let Some(value) = iter.next().map_err(js_error_to_postgres)? {
            if value.is_empty_or_undefined_or_null() {
                // Allow postgres to decide the type
                fields.push(0);
                name.extend_from_slice(b".null");
                continue;
            }

            let tag = tag_jsc::from_js(global_object, value).map_err(js_error_to_postgres)?;

            match tag {
                Tag::int8 => name.extend_from_slice(b".int8"),
                Tag::int4 => name.extend_from_slice(b".int4"),
                // Tag::int4_array => name.extend_from_slice(b".int4_array"),
                Tag::int2 => name.extend_from_slice(b".int2"),
                Tag::float8 => name.extend_from_slice(b".float8"),
                Tag::float4 => name.extend_from_slice(b".float4"),
                Tag::numeric => name.extend_from_slice(b".numeric"),
                Tag::json | Tag::jsonb => name.extend_from_slice(b".json"),
                Tag::bool => name.extend_from_slice(b".bool"),
                Tag::timestamp => name.extend_from_slice(b".timestamp"),
                Tag::timestamptz => name.extend_from_slice(b".timestamptz"),
                Tag::bytea => name.extend_from_slice(b".bytea"),
                _ => name.extend_from_slice(b".string"),
            }

            match tag {
                Tag::bool
                | Tag::int4
                | Tag::int8
                | Tag::float8
                | Tag::int2
                | Tag::numeric
                | Tag::float4
                | Tag::bytea => {
                    // We decide the type
                    fields.push(Int4::from(tag.0));
                }
                _ => {
                    // Allow postgres to decide the type
                    fields.push(0);
                }
            }
        }

        if iter.any_failed() {
            return Err(crate::Error::InvalidQueryBinding);
        }
        // max u64 length is 20, max prepared_statement_name length is 63
        let prepared_statement_name: Box<[u8]> = if unnamed {
            Box::default()
        } else {
            use std::io::Write;
            let mut v: Vec<u8> = Vec::new();
            write!(
                &mut v,
                "P{}${}",
                bstr::BStr::new(&name[..name.len().min(40)]),
                prepared_statement_id,
            )
            .expect("unreachable");
            v.into_boxed_slice()
        };

        Ok(Signature {
            prepared_statement_name,
            name: name.into_boxed_slice(),
            fields: fields.into_boxed_slice(),
            query: Box::<[u8]>::from(query),
        })
    }
}
